src/signing/stronghold.rs

```diff
@@ -1,4 +1,533 @@
+                    err
+                );
+                Err(crate::Error::StrongholdProcedureError(format!("{:?}", err)))
+            }
+        }
+    }
+
+    /// Execute [Procedure::Ed25519PublicKey] in Stronghold to get an Ed25519 public key from the SLIP-10 private key
+    /// located in `private_key`.
+    async fn ed25519_public_key(&self, private_key: Location) -> Result<[u8; 32]> {
+        match self
+            .stronghold
+            .runtime_exec(Procedure::Ed25519PublicKey { private_key })
+            .await
+        {
+            // Ed25519 public key get success.
+            ProcResult::Ed25519PublicKey(ResultMessage::Ok(pubkey)) => Ok(pubkey),
+            // Ed25519 public key get failure.
+            // XXX: Should we create a separate error type for this error?
+            ProcResult::Ed25519PublicKey(ResultMessage::Error(err)) => Err(crate::Error::StrongholdProcedureError(err)),
+            // Generic Stronghold procedure failure.
+            ProcResult::Error(err) => Err(crate::Error::StrongholdProcedureError(err)),
+            // Unexpected result type, which should never happen!
+            err => {
+                warn!(
+                    "StrongholdSigner::ed25519_public_key(): unexpected result from Stronghold: {:?}",
+                    err
+                );
+                Err(crate::Error::StrongholdProcedureError(format!("{:?}", err)))
+            }
+        }
+    }
+
+    /// Execute [Procedure::Ed25519Sign] in Stronghold to sign `msg` with `private_key` stored in the Stronghold vault.
+    async fn ed25519_sign(&self, private_key: Location, msg: &[u8]) -> Result<[u8; 64]> {
+        match self
+            .stronghold
+            .runtime_exec(Procedure::Ed25519Sign {
+                private_key,
+                msg: msg.to_vec(),
+            })
+            .await
+        {
+            // Ed25519 sign success.
+            ProcResult::Ed25519Sign(ResultMessage::Ok(msg)) => Ok(msg),
+            // Ed25519 sign failure.
+            // XXX: Should we create a separate error type for this error?
+            ProcResult::Ed25519Sign(ResultMessage::Error(err)) => Err(crate::Error::StrongholdProcedureError(err)),
+            // Generic Stronghold procedure failure.
+            ProcResult::Error(err) => Err(crate::Error::StrongholdProcedureError(err)),
+            // Unexpected result type, which should never happen!
+            err => {
+                warn!(
+                    "StrongholdSigner::ed25519_sign(): unexpected result from Stronghold: {:?}",
+                    err
+                );
+                Err(crate::Error::StrongholdProcedureError(format!("{:?}", err)))
+            }
+        }
+    }
+
+    /// Sign on `essence`, unlock `input` by returning a [UnlockBlock].
+    async fn signature_unlock(&self, input: &InputSigningData, essence: &TransactionEssence) -> Result<UnlockBlock> {
```

Comment:
We should accept the hashed essence TransactionEssence, so we only need to hash it once